data/transactions/logic/opcodeExplain.go

```diff
@@ -0,0 +1,332 @@
+// Copyright (C) 2019-2023 Algorand, Inc.
+// This file is part of go-algorand
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package logic
+
+import "github.com/algorand/go-algorand/data/basics"
+
+// debugStackExplain explains the effect of an opcode over the stack
+// with 2 integers: deletions and additions, representing pops and inserts.
+// An opcode may delete a few variables from stack, then add a few to stack.
+type debugStackExplain func(*EvalContext) (int, int)
+
+// AppStateOperationT stands for the operation enum to app state, should be one of create, write, read, delete.
+type AppStateOperationT uint64
+
+const (
+	// AppStateCreate stands for creating an app state.
+	AppStateCreate AppStateOperationT = 1 << iota
+
+	// AppStateWrite stands for writing to an app state.
+	AppStateWrite
+
+	// AppStateRead stands for reading from an app state.
+	AppStateRead
+
+	// AppStateDelete stands for deleting an app state.
+	AppStateDelete
+)
+
+// AppStateEnum stands for the enum of app state type, should be one of global/local/box.
+type AppStateEnum uint64
+
+const (
+	// GlobalState stands for global state of an app.
+	GlobalState AppStateEnum = iota + 1
+
+	// LocalState stands for local state of an app.
+	LocalState
+
+	// Box stands for box storage of an app.
+	Box
+)
+
+// stateChangeExplain explains how an opcode change the app's state with a quadruple:
+// AppStateEnum stands for which app state: local/global/box,
+// AppStateOperationT stands for read/write/create/delete/check-existence,
+// together with key for touched state
+type stateChangeExplain func(ctx *EvalContext) (AppStateEnum, AppStateOperationT, basics.AppIndex, basics.TealValue, string)
+
+func opPushIntsStackChange(cx *EvalContext) (deletions, additions int) {
+	// NOTE: WE ARE SWALLOWING THE ERROR HERE!
+	// FOR EVENTUALLY IT WOULD ERROR IN ASSEMBLY
+	intc, _, _ := parseIntImmArgs(cx.program, cx.pc+1)
+
+	additions = len(intc)
+	return
+}
```

Comment:
I wonder if we could get all of these explain functions tested by checking them against what happens in `TestReturnTypes` which is trying every instruction to see if it meets its prototype.

Reply:
It would be a bit involved to modify `TestReturnTypes`, for evaluating generating program give you `cx *EvalContext` **after** evaluating the opcode, while we are unsure what it is like before evaluation. I defer this testing a bit in initial state PR for thorough checks.